source/extensions/filters/listener/tls_inspector/tls_inspector.cc

```diff
@@ -228,14 +228,20 @@ std::vector<absl::string_view> Filter::getAlpnProtocols(const unsigned char* dat
     return protocols;
   }
 
-  absl::string_view str(reinterpret_cast<const char*>(data));
   for (unsigned int i = 0; i < len;) {
-    uint32_t protocol_length = 0;
-    protocol_length <<= 8;
-    protocol_length |= data[i];
-    ++i;
-    absl::string_view protocol(str.substr(i, protocol_length));
-    protocols.push_back(protocol);
+    uint8_t protocol_length = data[i++];
+
+    if (protocol_length == 0) {
+      ENVOY_LOG(debug, "tls inspector: ALPN protocol length is zero");
+      break;
+    }
+
+    if (protocol_length > (len - i)) {
+      ENVOY_LOG(debug, "tls inspector: ALPN protocol length is too long");
+      break;
+    }
+
+    protocols.emplace_back(reinterpret_cast<const char*>(data + i), protocol_length);
     i += protocol_length;
```

Comment:
readability nit: afaict this increment could move into the usual position in the for() loop on the premise that we declare it outside of that and initialize it to zero -- but that probably is just personal preference